src/miner.cpp

```diff
@@ -2188,6 +2188,8 @@ void static BitcoinMiner()
         if (minerThreads != NULL)
         {
             minerThreads->interrupt_all();
+            // std::cout << "Waiting for mining threads to stop..." << std::endl;
+            minerThreads->join_all();    // prevent thread overlapping   
```

Comment:
Given the lack of interruption points, using both interrupt_all() and join_all() can (and "randomly" will) bring everything to a grinding halt when between interruptibles and unjoinable.

Reply:
Good point about lack of interruption points. 
Currently we do not have join what leads to inaccurate thread shutdown.
And adding join_all works well in marmara and fixed a crash when a user called 'setgenerate false' and 'setgenerate true' quickly in a sequence.
But I missed an extra interruption_point (from the marmara code) in komodo_waituntilelegible which does a long loop which would make thread shutdowns faster, adding it

Reply:
Could you explain what issue you mean? I believe joins may create an issue when a thread can never be joined and hangs because it may have a long loop without interruption points or sleep or wait calls. If we have such loops we need to fix this to provide thread graceful shutdown to work.
(I am going to run my dev node on this branch for testing)

Reply:
`BitcoinMiner` only contains a single interruption_point()
That point comes **after** the while loop at:
https://github.com/KomodoPlatform/komodo/blob/dcbf657cabac274ad54f3528396ad9524636cefb/src/miner.cpp#L1969
which can hold up to 17.5 minutes, as set by:
https://github.com/KomodoPlatform/komodo/blob/dcbf657cabac274ad54f3528396ad9524636cefb/src/miner.cpp#L860
Ironically, the likelihood of that longest pause is increased by the new "stall reduction" code increasing the possibility for `gpucount` to reach 0 (assuming the rand hits 1056, which is possible).

Being async, the thought would be that this part would stick on its own and there would be no care; however, in reality, there are about a dozen circumstances where it locks everything (one of those circumstances being that threads are ignorant of each other and on advanced hardware come back with multiple solves when many miner threads are used).

When you look at what "should be" vs "what is", it shouldn't be a problem because NNs "should be" only running one thread; however, this is becoming decreasingly true with more and more NNs seeking to hit smaller and more predictable gaps.

Reply:
This is a good catch, this loop.
However I believe when a user calls setgenerate false and then setgenerate true without join there will be two running threads for some time and this is basically not good at all (in a staking chain this could even create a crash as there is a static komodo_staking *array var that could be corrupted in this case). So I think threads should be stopped gracefully by joining.
This loop you mentioned has sleep() inside and we can replace it on boost::this_thread::sleep_for function (which allows to interrupt the thread) and we should check other remaining loops in miner.cpp to add interruption points in a similar way

Reply:
> and we can replace it on boost::this_thread::sleep_for function (which allows to interrupt the thread) and we should check other remaining loops in miner.cpp to add interruption points in a similar way

However you want to do it, just want to make sure that you/everyone is aware that doing it as-is **will** lock up NNs; so, whatever way it's done to protect stakers that _could_ crash needs to be done in such a was as to protect both.

Reply:
I am not sure about your doubts though.
Deleting a thread object without join is an obvious bug IMO and should be fixed.
Fixing it may add some delay on daemon stopping or setgenerate false but this is not a lock-up if we do this properly.
Checking all loops in miner.cpp...

Reply:
Haven't tested it yet, but (along with the others)
https://github.com/KomodoPlatform/komodo/blob/12261b951ef040a725a6a8c2b69f0916f4470928/src/miner.cpp#L1989
does look like it'll solve the issue of my concern. ty